### PR TITLE
feat(consumer) validate credential secret fields

### LINKED ISSUE #532: Ingress Controller 0.7.1 Crashes when Kubernetes Secret is not formatted as expected
### Summary
Kong Ingress Controller Crashes when Kong Credential secret does not contain key `key`

**Kong Ingress controller version**
0.7.1

**Kong or Kong Enterprise version**
1.4.3

### Kubernetes version
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-13T11:51:44Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.9-eks-c0eccc", GitCommit:"c0eccca51d7500bb03b2f163dd8d534ffeb2f7a2", GitTreeState:"clean", BuildDate:"2019-12-22T23:14:11Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}

### Environment
AWS EKS

### What happened
Kong Ingress Controller Crashes trying to read in a secret that does not have a key in the secret named "key"

### Expected behavior
Kong Ingress Controller will gracefully handle this without null pointer exception resulting in crash

### Steps To Reproduce

1. Create a secret with key `kongCredType` and do not create a subsequent key named `key`

### Output
```
E0207 19:49:04.782376       1 runtime.go:73] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 277 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x142a600, 0x2229630)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/runtime/runtime.go:69 +0x7b
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/runtime/runtime.go:51 +0x82
panic(0x142a600, 0x2229630)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/hbagdi/deck/file.(*stateBuilder).ingestKeyAuths(0xc000ec99f0, 0xc00032b0a0, 0x2, 0x2, 0x2, 0xc00032b0a0)
	/go/pkg/mod/github.com/hbagdi/deck@v0.7.1-0.20191223190449-3d9f90945d9d/file/builder.go:222 +0x408
github.com/hbagdi/deck/file.(*stateBuilder).consumers(0xc000ec99f0)
	/go/pkg/mod/github.com/hbagdi/deck@v0.7.1-0.20191223190449-3d9f90945d9d/file/builder.go:161 +0x57b
github.com/hbagdi/deck/file.(*stateBuilder).build(0xc000ec99f0, 0x0, 0x0, 0x4)
	/go/pkg/mod/github.com/hbagdi/deck@v0.7.1-0.20191223190449-3d9f90945d9d/file/builder.go:56 +0x124
github.com/hbagdi/deck/file.Get(0xc0009e00e0, 0xc000049200, 0x1, 0x4, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/pkg/mod/github.com/hbagdi/deck@v0.7.1-0.20191223190449-3d9f90945d9d/file/reader.go:51 +0x130
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).onUpdateDBMode(0xc0009a6000, 0xc0009e00e0, 0x13c6960, 0xc000c7a8e0)
	/kong-ingress-controller/internal/ingress/controller/kong.go:161 +0x1c4
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).OnUpdate(0xc0009a6000, 0xc000650a80, 0x0, 0x0)
	/kong-ingress-controller/internal/ingress/controller/kong.go:63 +0x26c
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).syncIngress(0xc0009a6000, 0x149a360, 0xc000156820, 0xc017934d78, 0x2adca41b2681c)
	/kong-ingress-controller/internal/ingress/controller/controller.go:117 +0x1fa
github.com/kong/kubernetes-ingress-controller/internal/ingress/task.(*Queue).worker(0xc00046fce0)
	/kong-ingress-controller/internal/ingress/task/queue.go:111 +0x2df
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0005407a8)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000ec9fa8, 0x3b9aca00, 0x0, 0x1, 0xc00051e2a0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:88
github.com/kong/kubernetes-ingress-controller/internal/ingress/task.(*Queue).Run(0xc00046fce0, 0x3b9aca00, 0xc00051e2a0)
	/kong-ingress-controller/internal/ingress/task/queue.go:59 +0x6b
created by github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).Start
	/kong-ingress-controller/internal/ingress/controller/controller.go:232 +0xe8
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x11ab8c8]
goroutine 277 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/runtime/runtime.go:58 +0x105
panic(0x142a600, 0x2229630)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/hbagdi/deck/file.(*stateBuilder).ingestKeyAuths(0xc000ec99f0, 0xc00032b0a0, 0x2, 0x2, 0x2, 0xc00032b0a0)
	/go/pkg/mod/github.com/hbagdi/deck@v0.7.1-0.20191223190449-3d9f90945d9d/file/builder.go:222 +0x408
github.com/hbagdi/deck/file.(*stateBuilder).consumers(0xc000ec99f0)
	/go/pkg/mod/github.com/hbagdi/deck@v0.7.1-0.20191223190449-3d9f90945d9d/file/builder.go:161 +0x57b
github.com/hbagdi/deck/file.(*stateBuilder).build(0xc000ec99f0, 0x0, 0x0, 0x4)
	/go/pkg/mod/github.com/hbagdi/deck@v0.7.1-0.20191223190449-3d9f90945d9d/file/builder.go:56 +0x124
github.com/hbagdi/deck/file.Get(0xc0009e00e0, 0xc000049200, 0x1, 0x4, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/pkg/mod/github.com/hbagdi/deck@v0.7.1-0.20191223190449-3d9f90945d9d/file/reader.go:51 +0x130
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).onUpdateDBMode(0xc0009a6000, 0xc0009e00e0, 0x13c6960, 0xc000c7a8e0)
	/kong-ingress-controller/internal/ingress/controller/kong.go:161 +0x1c4
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).OnUpdate(0xc0009a6000, 0xc000650a80, 0x0, 0x0)
	/kong-ingress-controller/internal/ingress/controller/kong.go:63 +0x26c
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).syncIngress(0xc0009a6000, 0x149a360, 0xc000156820, 0xc017934d78, 0x2adca41b2681c)
	/kong-ingress-controller/internal/ingress/controller/controller.go:117 +0x1fa
github.com/kong/kubernetes-ingress-controller/internal/ingress/task.(*Queue).worker(0xc00046fce0)
	/kong-ingress-controller/internal/ingress/task/queue.go:111 +0x2df
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0005407a8)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000ec9fa8, 0x3b9aca00, 0x0, 0x1, 0xc00051e2a0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:88
github.com/kong/kubernetes-ingress-controller/internal/ingress/task.(*Queue).Run(0xc00046fce0, 0x3b9aca00, 0xc00051e2a0)
	/kong-ingress-controller/internal/ingress/task/queue.go:59 +0x6b
created by github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).Start
	/kong-ingress-controller/internal/ingress/controller/controller.go:232 +0xe8
```

### PR DESCRIPTION
**Pull request by rainest**
**What this PR does / why we need it**:

This is the band-aid solution discussed in https://github.com/Kong/deck/pull/223#issuecomment-719042303

It's effectively the same set of checks on the go-kong structs, but in the controller instead of deck. This allows us to log richer information than we would if we delegated this check to either deck or go-kong.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: Fix #532

**Special notes for your reviewer**:
Per comments in deck#223, this is an imperfect stopgap solution to a complicated problem

No new tests, as there's not a whole lot to test here. We could add one with one or more broken credentials, but given the band-aid nature of the solution (checks specific to each credential type, in a codebase that ideally shouldn't have them), it seems like they'd be of limited use. Regressions shouldn't occur unless we remove this code, and we shouldn't be doing that unless we're replacing it with some better system, which would have its own, more effective checks.

### CODE REVIEW

**Review by mflendrich — CHANGES_REQUESTED** [review]
Looks good but needs to be unit-tested.

**Comment by rainest** [comment]
@mflendrich per the original PR notes, do we indeed want unit tests for this? We can't unit test efficiently in the controller due to limitations elsewhere, and long-term we want to remove those along with this band-aid solution.

Ideally we want to get to something more generic ("give controller bad credential, controller discards bad credential, bubbles issue up to some layer that informs user via resource status") for any credential type, but we can't yet. Testing each individual known bad credential now does verify this, but seems rather futile long-term; it's test cruft that we'll eventually discard.

**Comment by mflendrich** [comment]
> We can't unit test efficiently in the controller due to limitations elsewhere

Wouldn't adding cases under https://github.com/Kong/kubernetes-ingress-controller/blob/99486c76b547dd0600272a5023a449d39734e00b/internal/ingress/controller/parser/kongstate/consumer_test.go#L11 work?

> and long-term we want to remove those along with this band-aid solution.

I think "this is only temporary" isn't a good reason to skip testing, because temporary solutions often tend to exceed their projected lifetime :)

> Ideally we want to get to something more generic ("give controller bad credential, controller discards bad credential, bubbles issue up to some layer that informs user via resource status") for any credential type, but we can't yet. Testing each individual known bad credential now does verify this, but seems rather futile long-term; it's test cruft that we'll eventually discard.

We can think of it this way: in this PR we add the statement "if the user provides a credential without identifier (username/client id/etc), KIC will drop it as invalid" to the contract between KIC and the user. We don't need to test "each individual known bad credential" - just credentials with the ID field missing.

**Comment by rainest** [comment]
It sort of works--the issue is that there's no universal missing ID that we can test. Every type of credential uses a different alternative key field, e.g. key-auth credentials use the key, basic-auth credentials use the username, oauth2 uses the client_id, etc. The structure of each check is identical, but the exact input is different.

For example, [test.diff.txt](https://github.com/Kong/kubernetes-ingress-controller/files/5527388/test.diff.txt) provides a valid test for key-auth, with a minor catch (more on that later--it's not relevant beyond my being bad at Golang), but it doesn't test the rest.

We could get around that by instead breaking the check into its own function that accepts the critical field and returns an error if it's nil, but IMO that strays a bit too far into tail-wagging-dog territory: we only add an additional function and call to make it easier to cover. That'd be easier to justify if the check were complex, but it isn't: an if block with an "equals nil?" condition that formats and returns an error if the condition is true is dead simple, to the point where I'd expect vet to reject it if there's any chance it'd fail.

I understand the concern about the workaround being longer-lived than we'd like, and indeed fear that it will remain here for quite some time myself, but I think there are good mitigating factors:

- The reason why we have the workaround is relatively well-understood.
- The workaround is simple, albeit repetitive.
- The workaround is accompanied by an explanatory comment explaining its purpose to future readers, and it instructs them not to touch it unless we've addressed the root cause.
- If we have addressed the root cause, any existing tests for the workaround should be removed and replaced by something that tests the larger fix. That fix hopefully _does_ allow us to write a single test, because it'd presumably delegate this check to deck or go-kong, and we'd cover it with a single bad input to `decodeCredential()` or similar.

I don't particularly like omitting tests, but I also don't particularly like clutter in the test cases. In light of that, I propose that we formalize https://github.com/Kong/deck/pull/223#issuecomment-718209730 into a research spike backlog item and then go with one of the following:

- Write a single test similar to the patch. Acknowledge that coverage is incomplete, but that we know all such checks are similar enough that testing each individually doesn't provide much benefit. Imperfect, but minimal clutter.
- Write tests for every check anyway.
- Write tests for none, acknowledging that we understand the risks but think they're mitigated.

Which of those do you prefer?

Anyway, onward with part two, in which I uncovered a new problem that I'm having trouble fixing. The attached test modifies the error printout to omit the consumer username, because that caused a segfault with the copied test parameters. The original code didn't do anything with the consumer username, so using completely empty consumers was fine. I'd wanted to address that with something like [invalid.diff.txt](https://github.com/Kong/kubernetes-ingress-controller/files/5527497/invalid.diff.txt), but that makes Go mad:

```
$ go vet ./...
# github.com/kong/kubernetes-ingress-controller/internal/ingress/controller/parser/kongstate
vet: ./consumer_test.go:57:5: too few values in struct literal
```
I understand it's unhappy about that because I didn't key the kong.Consumer, but unfortunately doing so is impossible, as that bit [has no key](https://github.com/Kong/kubernetes-ingress-controller/blob/1.0.0/internal/ingress/controller/parser/kongstate/consumer.go#L14). I could add one, but that means I can no longer use use `*c.Username` elsewhere. I like that shorthand :disappointed: 

Is there a way to construct the literal otherwise? Golang issue 9859 suggests you cannot, i.e. there's no way around the "cannot use promoted field Consumer.Username in struct literal of type Consumer" vet error you get if you try and add Username directly to the KIC Consumer literal.

**Comment by rainest** [comment]
With the Go syntax sorted, [test-all.diff.txt](https://github.com/Kong/kubernetes-ingress-controller/files/5533679/test-all.diff.txt) is the diff that checks all pending the question above.

As you can see, it's very repetitive, since we don't care about anything other than the critical field, so the credential is just completely blank always--the only thing that changes between the test cases is the credential type. IMO it's basically boilerplate clutter, but I'll add it if you prefer.

**Comment by mflendrich** [comment]
> It sort of works--the issue is that there's no universal missing ID that we can test. Every type of credential uses a different alternative key field, e.g. key-auth credentials use the key, basic-auth credentials use the username, oauth2 uses the client_id, etc. The structure of each check is identical, but the exact input is different.

This is unfortunate indeed, but this structure means that each error case is a separate logical path (see the code coverage printout for this PR - red lines are not covered by any automated test):

image

Since unit tests are typically expected to cover all meaningful code paths (and a code path rejecting an entity from further processing is meaningful beyond my doubt), we do need to cover them by tests. Here this unfortunately means having a test case for each auth type. But with table-based tests this is should be as simple as an array with 5 (or so) credentials, so repetition will be minimal (if any).

So let's have a table-based test with the case arary similar to this (it's possible that positive cases are satisfactorily covered by other tests already):

```go
{
  {cred: {a secret decoding to KeyAuth}, wantErr: true},
  {cred: {a secret decoding to KeyAuth}, wantErr: false},

  {cred: {a secret decoding to BasicAuth}, wantErr: true},
  {cred: {a secret decoding to BasicAuth}, wantErr: false},

  // ... and so on, for other credential types
}
```

**Comment by mflendrich** [comment]
> I think there are good mitigating factors

Let's have that discussion on zoom, if necessary - that's quicker than talking through GH comments.

I think that this way forward:
> Write tests for every check anyway.

should be as simple as writing a short test with one table-based case for each code path (that means pass/fail for each.

> I uncovered a new problem 

I recall that this is already resolved - right?

**Comment by rainest** [comment]
> It should be possible to reuse a single Consumer fixture for all cases.

Consumer inputs for the existing tests are identical, but outputs aren't. The positive cases test that the Consumers were modified correctly, e.g. the [input for one of the JWT cases](https://github.com/Kong/kubernetes-ingress-controller/blob/1.0.0/internal/ingress/controller/parser/kongstate/consumer_test.go#L197) and [its output](https://github.com/Kong/kubernetes-ingress-controller/blob/1.0.0/internal/ingress/controller/parser/kongstate/consumer_test.go#L204-L213). I suspect that they instantiate a new Consumer per test to account for that. The new tests in 346b0f8 (formerly the diff) follow that pattern, similar to the [one existing negative test](https://github.com/Kong/kubernetes-ingress-controller/blob/1.0.0/internal/ingress/controller/parser/kongstate/consumer_test.go#L24-L32).

IMO if we don't want to forsake coverage here then we should go ahead and accept that adding new tests that follow the existing structure means verbosity. The existing code here doesn't have the concept of a generic credential, which is why it needs separate paths to handle each separately.

**Review by mflendrich — COMMENTED** [review]
One comment, otherwise LGTM

File: internal/ingress/controller/parser/kongstate/consumer_test.go [hunk 1]
```diff
@@ -46,6 +47,24 @@ func TestConsumer_SetCredential(t *testing.T) {
...
 		},
+		{
+			name: "key-auth without key",
+			args: args{
+				credType: "key-auth",
+				consumer: &Consumer{
+					Consumer: kong.Consumer{
+						Username: &username,
+					},
+				},
```
Comment (mflendrich) [inline internal/ingress/controller/parser/kongstate/consumer_test.go:58]:
```
emptyConsumer := Consumer{
	Consumer: kong.Consumer{
		Username: &username,
	},
}
```

this way we could replace `consumer: ...` with `consumer: &emptyConsumer` (and the same for `result`, saving all the repetitions.

Comment (rainest) [reply internal/ingress/controller/parser/kongstate/consumer_test.go]: Left independent but squashed into a single line per literal as discussed.

Comment (mflendrich) [reply internal/ingress/controller/parser/kongstate/consumer_test.go]: As per discussion elsewhere: we remain with literals in every test case but roll them up in a single line.

**Review by mflendrich — APPROVED** [review]

**Pull request merged**